src/util/gamma_picker.cpp

```diff
@@ -84,17 +89,27 @@ namespace lws
     static_assert(std::is_empty<crypto::random_device>(), "random_device is no longer cheap to construct");
     static constexpr const crypto::random_device engine{};
     const auto end = offsets().end() - CRYPTONOTE_DEFAULT_TX_SPENDABLE_AGE;
+    const uint64_t num_rct_outputs = spendable_upper_bound();
 
     for (unsigned tries = 0; tries < 100; ++tries)
     {
-      std::uint64_t output_index = std::exp(gamma(engine)) * outputs_per_second;
-      if (offsets().back() <= output_index)
+      double output_age_in_seconds = std::exp(gamma(engine));
+
+      // match wallet2 implementation to avoid creating distinct anonymity pools
+      // shift output back by unlock time to apply distribution from chain tip
+      if (output_age_in_seconds > defaut_unlock_time)
+        output_age_in_seconds -= defaut_unlock_time;
```

Comment:
Looks like I need to go back over the review that just occurred in `wallet2`.

This is creating a bias in the selection algorithm. It is now more likely to select blocks 10-20.

Reply:
Sorry let me restate -

Looks like I will have to go back over `wallet2` PR description more closely. I don't see how this `if` / `else` logic is correct, this is deviating from the source paper a bit.

Reply:
Ok, just read the justification for the `recent_spend_window`. It probably makes sense given that the current `continue` creates a bias that doesn't follow the original intent of the source paper.

And I see what did you this logic, but I would like you to refactor. The existing version was already selecting from the chain tip, but this adds two subtractions (here and line 109) to match the code in `wallet2`. Trying to match that code led to an unnecessary `throw`.

Revert everything, and change the continue on the current line 97:
```c++
if (selection == end)
  output_index = *(offsets.rbegin() + crypto::rand_range(CRYPTONOTE_DEFAULT_TX_SPENDABLE_AGE, std::min(50, offsets.size() - 1)))
static_assert(CRYPTONOTE_DEFAULT_TX_SPENDABLE_AGE <= 50, "spendable must be less than spend window");
```

Reply:
See Github for fixed code snippet

Reply:
On the subtraction: I figured it makes sense *not* to consider locked outputs when selecting outputs at all. Why should the outputs in that 10 block window be considered in the output line? There is no feasible way for a user to spend those outputs, so why should it be the case that they should have any impact at all on what outputs are likely to be spent by a user?  Imagine outputs spike in the 10 block window from the tip. This would cause the algorithm to select these outputs in greater frequency (and then redraw), but there is no reason for the algorithm to select those outputs in the first place because those outputs are not even feasibly spendable outputs. The general issue here is that the paper was written before the 10 block unlock time was enforced by consensus, so there needs to be some deviation to factor that in. This approach seemed to be the most sensible to me.

The above reasoning is why I stuck with moo's original approach where `num_rct_outputs` is the start of the output line where you ultimately subtract from, and went with the fairly odd (and admittedly code smelly) 2-step subtraction approach. 

On the throw: It's unnecessary because it should never happen. I just kept it there as a sanity check to signal the intent.

I think whatever we settle on here should be applied to wallet2 as well. So if we decide against the subtraction approach, we should probably make the change in wallet2 to match this implementation.

I should've asked you to review 7821 closely too, it would've been nice to have your eyes on it then as well and have this discussion then. I originally wanted to go with the exact same approach as yours (before I saw this code) -- your approach was the fix I initially proposed to moo when I initially reported it too. I wished the cleaner solution was also the best solution, but it didn't seem that way to me after sitting with it for a long time.

Reply:
Random over `output_age_in_seconds` has a different distribution than random over `blocks` (I think that was your point).

In that case, rename `num_rct_outputs` to `num_unlocked_rct_outputs`. At first glance I thought this was shifting the distribution to the unlocked outputs. Also make `std::runtime_error` -> `std::logic_error` with some message indicating that it shouldn't happen (i.e. the logic needs to be looked at again).